numba/parfors/array_analysis.py

```diff
@@ -2064,6 +2064,17 @@ def _analyze_op_arrayexpr(self, scope, equiv_set, expr):
         )
 
     def _analyze_op_build_tuple(self, scope, equiv_set, expr):
+        # For the moment, we can't do anything with tuples that
+        # contain arrays, compared to array dimensions.  Return
+        # None to say we won't track this tuple if a part of it
+        # is an array.
+        for x in expr.items:
+            if (
+                isinstance(x, ir.Var)
+                and isinstance(self.typemap[x.name], types.ArrayCompatible)
```

Comment:
Should this check include `types.BaseTuple` as well since that would cause the same nested shape situation?
Also, how hard would supporting nested shapes be?

Reply:
@stuartarchibald @ehsantn Below is the unique set of types that array analysis processes for build_tuple due to test_parfors.py.  The 2D array ones come from the test I just added for this PR.  Evidently, we were already seeing 1D arrays and through luck they weren't causing a problem.  Stepping back and asking how we use array analysis, we use it to see if two array dimensions, possibly propagated through tuples, are known to be the same size.  Thus, we could argue that the only useful type to track is some variety of integer.  I haven't given it a ton of thought but nested tuples don't seem to easily fit into the current array analysis approach and indeed it would be easy to hide potential associations from array analysis now by first putting them into nested tuples and then extracting from them.  I think the right thing to do going forward would be to take an inclusionary approach (because how would you possibly come up with a complete exclusionary list) where we only include things of integer type in array analysis.  That's going to be a much bigger change but to get around this issue for the short term I chose just to exclude the problematic type we were actually seeing.

[]
[array(bool, 1d, C), int64]
[array(bool, 1d, C), Literal[int](1)]
[array(bool, 1d, C), slice<a:b>]
[array(bool, 1d, C), uint64]
[array(float64, 2d, C)]
[array(float64, 2d, C), array(float64, 2d, C)]
[array(int64, 1d, C), Literal[int](4)]
[int64]
[int64, int64]
[int64, int64, int64]
[int64, int64, Literal[int](1)]
[int64, Literal[int](3)]
[int64, slice<a:b>]
[Literal[int](0), Literal[int](0)]
[Literal[int](0), slice<a:b>]
[Literal[int](0), uint64]
[Literal[int](1)]
[Literal[int](10), Literal[int](10)]
[Literal[int](10), Literal[int](12)]
[Literal[int](13), Literal[int](17)]
[Literal[int](13), Literal[int](17)]
[Literal[int](17), Literal[int](5)]
[Literal[int](1), int64, int64]
[Literal[int](1), int64, int64, Literal[int](1)]
[Literal[int](1), Literal[int](1)]
[Literal[int](1), Literal[int](2)]
[Literal[int](20), Literal[int](20)]
[Literal[int](2), Literal[int](3)]
[Literal[int](3), Literal[int](2), Literal[int](7)]
[Literal[int](3), Literal[int](4)]
[slice<a:b>, array(int64, 1d, C)]
[slice<a:b>, Literal[int](0)]
[slice<a:b>, slice<a:b>, Literal[int](0)]
[uint64, int64]
[uint64, Literal[int](0)]
[uint64, slice<a:b>]
[uint64, uint64]
[uint64, uint64, uint64]
[uint64, uint64, uint64, uint64]

Reply:
To know array dimension equivalence, array analysis needs to keep track of both integers and arrays (as has been the case for a while). The current code works for 1D arrays since they can be represented by a single equivalence class. The right approach is to support nested/multi-dimensional cases properly (which I believe is not that difficult due to experience with similar analysis).